validation-test/stdlib/Arrays.swift.gyb

```diff
@@ -138,12 +144,435 @@ ArrayTestSuite.test("${Kind}/popLast") {
 }
 % end
 
-public struct ArrayIndexTest<T: Collection> {
-  public enum RangeBox {
-    case closed(ClosedRange<Int>)
-    case halfClosed(Range<Int>)
+//===----------------------------------------------------------------------===//
+// COW(🐄) tests
+//===----------------------------------------------------------------------===//
+
+class COWBox<T: Equatable & CustomStringConvertible>: Equatable, CustomStringConvertible {
+  var value: T
+  
+  init(_ value: T) {
+    self.value = value
+  }
+  
+  var description: String {
+    return "Boxed: \(value.description)"
   }
   
+  static func ==(lhs: COWBox, rhs: COWBox) -> Bool {
+    return lhs.value == rhs.value
+  }
+}
+
+ArrayTestSuite.test("COW.Smoke") {
+  var a1 = Array<COWBox<Int>>(repeating: COWBox(0), count: 10)
+  var identity1 = a1._rawIdentifier()
+  
+  a1[0] = COWBox(1)
+  a1[1] = COWBox(2)
+  a1[2] = COWBox(3)
+  
+  var a2 = a1
+  expectEqual(identity1, a2._rawIdentifier())
+  
+  a2[3] = COWBox(4)
+  expectNotEqual(identity1, a2._rawIdentifier())
+  
+  a1[4] = COWBox(5)
+  expectEqual(identity1, a1._rawIdentifier())
+  
+  _fixLifetime(a1)
+  _fixLifetime(a2)  
+}
+
+func getCOWFastArray() -> Array<Int> {
+  var a = Array<Int>()
+  a.reserveCapacity(10)
+  a.append(1)
+  a.append(2)
+  a.append(3)
+  return a
+}
+
+func getCOWSlowArray() -> Array<COWBox<Int>> {
+  var a = Array<COWBox<Int>>()
+  a.reserveCapacity(10)
+  a.append(COWBox(1))
+  a.append(COWBox(2))
+  a.append(COWBox(3))
+  return a
+}
+
+ArrayTestSuite.test("COW.Fast.IndexesDontAffectUniquenessCheck") {
+  var a = getCOWFastArray()
+  let identity1 = a._rawIdentifier()
+  
+  var startIndex = a.startIndex
+  var endIndex = a.endIndex
+  assert(startIndex != endIndex)
+  assert(startIndex < endIndex)
+  assert(startIndex <= endIndex)
+  assert(!(startIndex >= endIndex))
+  assert(!(startIndex > endIndex))
+  
+  assert(identity1 == a._rawIdentifier())
+  
+  a[0] = 4
+  assert(identity1 == a._rawIdentifier())
+  
+  // Keep indexes alive during the calls above
+  _fixLifetime(startIndex)
+  _fixLifetime(endIndex)
+}
+
+ArrayTestSuite.test("COW.Slow.IndexesDontAffectUniquenessCheck") {
+  var a = getCOWSlowArray()
+  let identity1 = a._rawIdentifier()
+  
+  var startIndex = a.startIndex
+  var endIndex = a.endIndex
+  assert(startIndex != endIndex)
+  assert(startIndex < endIndex)
+  assert(startIndex <= endIndex)
+  assert(!(startIndex >= endIndex))
+  assert(!(startIndex > endIndex))
+  
+  assert(identity1 == a._rawIdentifier())
+  
+  a[0] = COWBox(4)
+  assert(identity1 == a._rawIdentifier())
+  
+  // Keep indexes alive during the calls above
+  _fixLifetime(startIndex)
```

Comment:
For the tests, I think `_blackHole` is a better choice than `_fixLifetime`, it will be completely un-breakable by the optimizer.